packages/flutter/lib/src/painting/image_cache.dart

```diff
@@ -247,6 +269,110 @@ class ImageCache {
+    this.weak = false,
+  }) : assert(!pending || !strong);
+
+  /// An image that has been submitted to [FlutterImageCache.putIfAbsent], but
+  /// not yet completed.
+  final bool pending;
+
+  /// An image that has been submitted to [FlutterImageCache.putIfAbsent], has
+  /// completed, fits based on the sizing rules of the cache, and has not been
+  /// evicted.
+  final bool strong;
+
+  /// An image that has been submitted to [FlutterImageCache.putIfAbsent], has
+  /// completed, and has at least one listener on its [ImageStreamCompleter].
+  final bool weak;
+
+  /// An image that either has not been submitted to
+  /// [FlutterImageCache.putIfAbsent] or has otherwise been evicted from the
+  /// [strong] and [weak] caches.
+  bool get untracked => !pending && !strong && !weak;
+
+  @override
+  bool operator ==(Object other) {
+    if (other.runtimeType != runtimeType) {
+      return false;
+    }
+    return other is ImageCacheLocation
+        && other.pending == pending
+        && other.strong == strong
+        && other.weak == weak;
+  }
+
+  @override
+  int get hashCode => hashValues(pending, strong, weak);
+}
+
+/// An [ImageCache] that adds weak tracking of images that have at least one
+/// listener on their [ImageStreamCompleter].
+///
+/// This cache also exposes internal dimensions of the implementation, such as
+/// checking whether an image for a given key is pending, weak, or completed.
+/// See [ImageCacheLocation] for more details.
+class FlutterImageCache extends ImageCache {
+  final Map<Object, ImageStreamCompleter> _weakCache = <Object, ImageStreamCompleter>{};
+
+  @override
+  ImageStreamCompleter _checkOtherResults(Object key) {
+    // We are weakly holding the image until there are no listeners.
+    // This does not respect the size parameters of the cache, because this
+    // indicates these images are visible on the screen or held by someone else.
+    return _weakCache[key];
+  }
+
+  @override
+  void _onImageStreamCompleterCreated(Object key, ImageStreamCompleter completer) {
+    _weakCache[key] = completer;
+    completer.addOnLastListenerRemovedCallback(() {
+      _weakCache.remove(key);
+    });
+  }
+
+  /// The [ImageCacheLocation] information for the given `key`.
+  ImageCacheLocation locationForKey(Object key) => ImageCacheLocation._(
+    pending: _pendingImages.containsKey(key),
+    strong: _cache.containsKey(key),
+    weak: _weakCache.containsKey(key),
+  );
+
+  /// The number of images being weakly held by the [ImageCache]. Compare with
+  /// [ImageCache.currentSize] for strongly held images.
+  int get weakImageCount => _weakCache.length;
+
+  /// The number of images being tracked as pending in the [ImageCache]. Compare
+  /// with [ImageCache.currentSize] for strongly held images.
+  int get pendingImageCount => _pendingImages.length;
+
+  /// Clears any weak references to images in this cache.
+  ///
+  /// This should not ordinarily be called by applications.
```

Comment:
Rather than saying it shouldn't be called, say when it _should_ be called.

Reply:
Done